File: JavaScript/DataStructures/array/customArray/myArray.js


class MyArray {

    constructor() {
      this.length = 0;
      this.data = {}
        
    };

    // O(1);
    get(index){
        return this.data[index];
    }
    // O(1);
    push(item){
        this.data[this.length] = item;
        this.length++;
        return this.length;
    }
    // O(1);
    pop(){
        delete this.data[this.length-1];
        this.length--;
    }
    // O(n);
    delete(index){
        const item = this.data[index];
        this.shiftItems(index);
        return item;
    }

    shiftItems(index){
        for(let i = index; i < this.length -1; i++){
            this.data[i] = this.data[i+1];
        }
        delete this.data[this.length -1];
        this.length--;
    }

}



const array = new MyArray();
array.push('Hi')
array.push('Toto')
array.push('Manco');
array.push('x2')

console.log(array);

array.pop();
console.log(array);

array.delete(1);
console.log(array);

File: Java/src/DataStructures/Arrays/customArray/MyArray.java
package DataStructures.Arrays.customArray;

public class MyArray {

    int length;
    int [] data;

    public MyArray() {
        this.length = 0;
        this.data = new int[5];
    }

    public int getIndex(int index) {
        return this.data[index];
    }

    public void push(int item){
        this.data[this.length] = item;
        length++;
    }

    public void pop(){
        int lastItem = this.data[this.length-1];
        this.data[lastItem] = 0;
        this.length--;
    }

    public void delete(int index){
        for (int i = index; i < this.length -1 ; i++) {
            this.data[i] = this.data[i+1];
        }

    }
}

class Main{
    public static void main(String[] args) {

        MyArray myArray = new MyArray();
        System.out.println(myArray.getIndex(0));
        System.out.println(myArray.length);
        myArray.push(1);
        myArray.push(2);
        myArray.push(3);
        myArray.push(4);
        System.out.println(myArray.length);
        System.out.println(myArray.getIndex(4));
        myArray.pop();
        System.out.println(myArray.length);
        myArray.delete(1);
        System.out.println(myArray.getIndex(2));
        System.out.println(myArray.getIndex(1));





    }
}
